velox/docs/develop/aggregate-functions.rst

```diff
@@ -583,27 +583,293 @@ You can see the documentation for all functions at :doc:`../functions/presto/agg
+   E20230714 14:57:33.734416 975289 HashStringAllocator.h:497] allocate 8
+   E20230714 14:57:33.734445 975289 HashStringAllocator.h:506] free 4
+   E20230714 14:57:33.734481 975289 HashStringAllocator.h:497] allocate 16
+   E20230714 14:57:33.734513 975289 HashStringAllocator.h:506] free 8
+   E20230714 14:57:33.734544 975289 HashStringAllocator.h:497] allocate 32
+   E20230714 14:57:33.734575 975289 HashStringAllocator.h:506] free 16
+   E20230714 14:57:33.734606 975289 HashStringAllocator.h:497] allocate 64
+   E20230714 14:57:33.734637 975289 HashStringAllocator.h:506] free 32
+   E20230714 14:57:33.734668 975289 HashStringAllocator.h:497] allocate 128
+   E20230714 14:57:33.734699 975289 HashStringAllocator.h:506] free 64
+   E20230714 14:57:33.734731 975289 HashStringAllocator.h:506] free 128
+
+
+Reallocating memory and copying data is not cheap. To avoid this overhead we
+introduced ValueList primitive and used it to implement array_agg.
+
+ValueList is an append-only data structure that allows appending values from any
+Velox Vector and reading values back into a Velox Vector. ValueList doesn’t
+require a contiguous chunk of memory and therefore doesn’t need to re-allocate
+and copy when it runs out of space. It just allocates another chunk and starts
+filling that up.
+
+ValueList is designed to work with data that comes from Velox Vectors, hence,
+its API is different from std::vector. You append values from a DecodedVector
+and read values back into a flat vector. Here is an example of usage:
+
+.. code-block:: c++
+
+   DecodedVector decoded(*data);
+
+   // Store data.
+   ValueList values;
+   for (auto i = 0; i < 100; ++i) {
+    values.appendValue(decoded, i, allocator());
+   }
+
+
+   // Read data back.
+   auto copy = BaseVector::create(DOUBLE(), 100, pool());
+   aggregate::ValueListReader reader(values);
+   for (auto i = 0; i < 100; ++i) {
+    reader.next(*copy, i);
+   }
+
+ValueList supports all types, so you can use it to append fixed-width values as
+well as strings, arrays, maps and structs.
+
+When storing complex types, ValueList serializes the values using
+ContainerRowSerde.
+
+ValueList preserves the null flags as well, so you can store a list of nullable
+values in it.
+
+The array_agg is implemented using ValueList for the accumulator.
+
+ValueList needs a pointer to the arena for appending data. It doesn’t take an
+arena in the constructor and doesn’t store it, because that would require 8
+bytes of memory per group in the aggregation operator. Instead,
+ValueList::appendValue method takes a pointer to the arena as an argument.
+Consequently, ValueList’s destructor cannot release the memory back to the
+arena and requires the user to explicitly call the free
+(HashStringAllocator*) method.
+
+min, max, and SingleValueAccumulator
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:func:`min` and :func:`max` functions store a single value in the accumulator
```

Comment:
maybe explicitly call that primitive types don't need this (and how they are handled instead)